### PR TITLE
[IMP] link_tracker: make different trackers from labels

### PR DESCRIPTION
**Pull request by flch-odoo**
Purpose: Independently track clicks resolving to the same url from
different links on the same mailing if associated with a different label.

Note that as we add a criterion in the unicity constraint, we are adding a
dimension that wasn't used before. Therefore, no problem should arise in
existing databases.

Expanding the regex pattern matching lead to much less readable code.

Ill-formed urls are updated in the mailing templates.

Tests are included and their formatting is condensed and uniformized.

Task-2788965

### CODE REVIEW

File: odoo/tools/mail.py [hunk 1]
```diff
@@ -259,7 +259,17 @@ def html_sanitize(src, silent=True, sanitize_tags=True, sanitize_attributes=Fals
...
 # retrieve inner content of the link
-HTML_TAG_URL_REGEX = URL_REGEX + r'([^<>]*>([^<>]+)<\/)?'
+TEXT_BEFORE_CLOSING_TAG_REGEX = r'([^<>]*>([^<>]+)<\/)?'
+NO_TAG = r'[^<>]*'
+QUOTE = r'[\'\"]'
+NO_QUOTE = r'[^\'\"]'
+
+HTML_TAG_URL_REGEX = URL_REGEX + TEXT_BEFORE_CLOSING_TAG_REGEX
+HTML_TAG_IMAGE_ALT_REGEX = (
+        URL_REGEX
```
Comment (std-odoo) [inline odoo/tools/mail.py:269]: the regex need to be "multiline" (re.ALL I think), otherwise this won't work
```html
<a href="https://www.odoo.com">
    <img src="https://www.odoo.com/logo.png" alt="image"/>
</a>
```

Also if there's space between the end of `<a>` and the start of `` the label is not catched
```html
<a href="https://www.odoo.com"> <img src="https://www.odoo.com/logo.png" alt="image"/> </a>
```

(+ add test for the first scenario)

**Review by std-odoo — COMMENTED** [review]
@flch-odoo Hi :) 

I added some comments, I think we can improve the regex parsing without too much cost,

Good job :)

File: odoo/tools/mail.py (hunk 1 above)
Comment (flch-odoo) [reply odoo/tools/mail.py]: Test added! All content except tags will work before the `<img` though, even on different lines without the flag as we don't use `^` or `$`.

File: odoo/tools/mail.py [hunk 2]
```diff
@@ -259,7 +259,17 @@ def html_sanitize(src, silent=True, sanitize_tags=True, sanitize_attributes=Fals
...
+        + rf'({NO_TAG}>{NO_TAG}\<img{NO_TAG}alt={QUOTE}({NO_QUOTE}+){QUOTE})?'
```
[the remark this answers is not in the document]
Comment (flch-odoo) [reply odoo/tools/mail.py]: Good points! I think got all those covered in tests too now :-)

**Review by flch-odoo — COMMENTED** [review]
Hello @std-odoo ,
Thanks for pushing me to make it more robust!

File: addons/link_tracker/models/mail_render_mixin.py [hunk 3]
```diff
@@ -36,22 +36,22 @@ def _shorten_links(self, html, link_tracker_vals, blacklist=None, base_url=None)
...
+        for match in set(re.findall(tools.HTML_TAG_URL_WITH_IMAGE_ALT_REGEX, html)):
+            long_url = match[2]
             # Don't shorten already-shortened links
             if long_url.startswith(short_schema):
                 continue
             # Don't shorten urls present in blacklist (aka to skip list)
             if blacklist and any(s in long_url for s in blacklist):
                 continue
-            label = (match[3] or '').strip()
+            label = (match[8] or match[6] or match[5]).strip()
```
Comment (std-odoo) [inline addons/link_tracker/models/mail_render_mixin.py:47]: to be sure it will never crash (not 100% sure it can crash :thinking: )
```suggestion
            label = (match[8] or match[6] or match[5] or '').strip()
```

File: addons/link_tracker/models/link_tracker.py [hunk 4]
```diff
@@ -132,7 +132,7 @@ def _get_title_from_url(self, url):
 
         return title
 
-    @api.constrains('url', 'campaign_id', 'medium_id', 'source_id')
+    @api.constrains('url', 'campaign_id', 'medium_id', 'source_id', 'label')
```
Comment (std-odoo) [inline addons/link_tracker/models/link_tracker.py:135]: shouldn't you make a migration script to ensure no "duplicated" link trackers exists ? (because you update the constraint)

**Review by std-odoo — COMMENTED** [review]
@flch-odoo Some last picky remarks :)

**Review by flch-odoo — COMMENTED** [review]
Thanks again, @std-odoo . Less changes for this round, though ;). I did update the test for triple-quoted strings, I am not seeing it more readable but I know I'm deep into it so biased.
Please let me know if something is still unclear.

File: addons/link_tracker/models/link_tracker.py (hunk 4 above)
Comment (std-odoo) [reply addons/link_tracker/models/link_tracker.py:135]: ah ok indeed :)

File: addons/link_tracker/tests/test_mail_render_mixin.py [hunk 5]
```diff
@@ -104,8 +102,52 @@ def test_shorten_links_html_including_base_url(self):
...
+<a href="https://www.odoo.com">there</a>, and in this
+<a href="https://www.odoo.com"><img src="https://www.odoo.com/logo.png" alt="image"/></a>
+and also <a href="https://www.odoo.com">
+    <img src="https://www.odoo.com/logo.png" fakealt="image3" alt="image2's trouble"></img>
+</a>
+Single/Nested quotes are not <a href="https://www.odoo.com"><img src='https://www.odoo.com/logo.png' alt='"scary"'/></a>
+Nor escaped <a href="https://www.odoo.com"> blurp <img src="https://www.odoo.com/logo.png" alt="ins \' ide"></a>
+        """.format(base_url=self.base_url)
+
+        expected_pattern = re.compile("""
```
Comment (std-odoo) [inline addons/link_tracker/tests/test_mail_render_mixin.py:118]: here you can use `r` and remove double escaping
```suggestion
        expected_pattern = re.compile(r"""
```

File: odoo/tools/mail.py [hunk 6]
```diff
@@ -259,10 +259,17 @@ def html_sanitize(src, silent=True, sanitize_tags=True, sanitize_attributes=Fals
...
-HTML_NEWLINES_REGEX = re.compile('<(div|p|br|tr)[^>]*>|\n')
-
+TEXT_BEFORE_CLOSING_TAG_REGEX = r'([^<>]*>([^<>]+)</)?'
+HTML_TAG_URL_REGEX = re.compile(URL_REGEX + TEXT_BEFORE_CLOSING_TAG_REGEX)
+HTML_TAGS_REGEX = re.compile(r'<.*?>')
+HTML_NEWLINES_REGEX = re.compile(r'<(div|p|br|tr)[^>]*>|\n')
+NO_TAG = r'[^<>]*'
+QUOTED_TEXT = r'("([^"]*)"|\'([^\']*)\')'
+HTML_TAG_URL_WITH_IMAGE_ALT_REGEX = re.compile(
+    rf'({URL_REGEX}'
```
[the remark this answers is not in the document]
Comment (std-odoo) [reply odoo/tools/mail.py:269]: ah ok, because you use findall and not group(0) right ?
ok then

**Review by std-odoo — COMMENTED** [review]
@flch-odoo Hi :) 

One last comment, otherwise LGTM :)

File: addons/link_tracker/tests/test_mail_render_mixin.py (hunk 5 above)
Comment (flch-odoo) [reply addons/link_tracker/tests/test_mail_render_mixin.py]: I applied this in these tests and updated all to use it (and f-formatting), which makes a slightly bigger diff but does reduce the final amount of code :-)

File: odoo/tools/mail.py [hunk 7]
```diff
@@ -261,8 +261,25 @@ def html_sanitize(src, silent=True, sanitize_tags=True, sanitize_attributes=Fals
...
 TEXT_URL_REGEX = r'https?://[\w@:%.+&~#=/-]+(?:\?\S+)?'
 # retrieve inner content of the link
-HTML_TAG_URL_REGEX = URL_REGEX + r'([^<>]*>([^<>]+)<\/)?'
+TEXT_BEFORE_CLOSING_TAG_REGEX = r'([^<>]*>([^<>]+)</)?'
+HTML_TAG_URL_REGEX = re.compile(URL_REGEX + TEXT_BEFORE_CLOSING_TAG_REGEX)
 HTML_TAGS_REGEX = re.compile('<.*?>')
+
+# Assumes that an absolute link does not start with `/`, `#`, or `.` and has 2+ characters.
+# Difficult to do more in order to support different protocols, local URIs, etc.
+URL_ABSOLUTE_REGEX = URL_REGEX.replace(r'([^\'"]+)', r'([^\'"#./][^\'"]+)')
```
Comment (tde-banana-odoo) [inline odoo/tools/mail.py:270]: Hein ?

**Review by tde-banana-odoo — COMMENTED** [review]
Have to say I don't understand anything. Probably linked to the burger I ate this noon.

File: odoo/tools/mail.py [hunk 8]
```diff
@@ -261,8 +261,25 @@ def html_sanitize(src, silent=True, sanitize_tags=True, sanitize_attributes=Fals
...
+NO_TAG_LIMIT = r'[^<>]*'
+ABSOLUTE_SRC_CAPTURE_REGEX = r'''\bsrc=["']([\w@:%.+&~#=/-]+)["']'''
+QUOTED_NON_EMPTY_TEXT = r'("([^"]+)"|\'([^\']+)\')'
+# See web_editor's convert_inline.js classToStyle
+_MAGIC_TAGS = r'''<p class=["']o_outlook_hack["'][^>]*>|<!--\[if mso]>'''
+HTML_TAG_URL_WITH_OPTIONAL_IMAGE_REGEX = re.compile(
```
[the remark this answers is not in the document]
Comment (flch-odoo) [reply odoo/tools/mail.py:276]: You could look at the whole bunch of tests to see why there is all that, otherwise I'll come to you on Wed.

[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply odoo/tools/mail.py:276]: Meh, tests are quite complex as most of them are high-end tests (aka on a complex html code). Could be great to add on each regex, a short sentence like ``find links with image: group returns (base_url, url, img_label)``

File: addons/link_tracker/tests/test_mail_render_mixin.py [hunk 9]
```diff
@@ -69,61 +69,110 @@ def test_shorten_links_html_skip_shorts(self):
...
+And a last, more complex: <a href="{self.base_url}/r/(\w+)">There!</a>""")
+
         new_content = self.env["mail.render.mixin"]._shorten_links(content, {})
 
         self.assertRegex(new_content, expected_pattern)
         matches = expected_pattern.search(new_content).groups()
         # 3rd and 4th lead to the same short_url
         self.assertEqual(matches[0], matches[1])
-        # 5th has different label but should currently lead to the same link
-        self.assertEqual(matches[1], matches[2])
```
Comment (tde-banana-odoo) [inline addons/link_tracker/tests/test_mail_render_mixin.py:108]: Shouldn't we check the other results are effectively different by the way ?

**Review by tde-banana-odoo — COMMENTED** [review]
Zboing, would say that it lacks comments for any review anyway. Could be great to have a clear docstring on each regex, with: what it does, what are the returned groups. That way we understand it without having to retro-engineer tests. Regexes are generally clear only for the people who wrote it ^^ .

**Review by flch-odoo — COMMENTED** [review]
Updated as per your suggestions, @tde-banana-odoo !

File: odoo/tools/mail.py [hunk 10]
```diff
@@ -283,14 +283,40 @@ def sanitize_handler(doc):
...
 # ----------------------------------------------------------
-# HTML/Text management
+# HTML/Text management -- See tests in base/test/test_mail.py for details
 # ----------------------------------------------------------
 
 URL_REGEX = r'(\bhref=[\'"](?!mailto:|tel:|sms:)([^\'"]+)[\'"])'
 TEXT_URL_REGEX = r'https?://[\w@:%.+&~#=/-]+(?:\?\S+)?'
 # retrieve inner content of the link
-HTML_TAG_URL_REGEX = URL_REGEX + r'([^<>]*>([^<>]+)<\/)?'
+TEXT_BEFORE_CLOSING_TAG_REGEX = r'([^<>]*>([^<>]+)</)?'
```
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/tools/mail.py:292]: Going even further, do we need that last `?` here? Cannot it be moved inside `HTML_TAG_URL_WITH_OPTIONAL_IMAGE_REGEX`
```suggestion
TEXT_BEFORE_CLOSING_TAG_REGEX = r'[^<>]*>([^<>]+)</'
```

**Review by Julien00859 — COMMENTED** [review]
HTMLLinkLabelConverter is much better than regexps, thanks!

**Comment by tde-banana-odoo** [comment]
Youpie a @Julien00859 ! And less regex. I will finally be able to review it then.

**Comment by flch-odoo** [comment]
Or here, @tde-banana-odoo :-) ?

File: addons/link_tracker/tools/html_link_label_converter.py [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+from collections.abc import Callable
+
+import lxml
+from markupsafe import Markup
+
+MAX_LABEL_LENGTH = 40  # arbitrary
+
+
+class HTMLLinkLabelConverter:
```
Comment (tde-banana-odoo) [inline addons/link_tracker/tools/html_link_label_converter.py:11]: Have to say I am not a fan of this kind of implementation ... it makes the code hard to follow: you define functions (with Odoo code as there is a "create" a bit hidden in the 'map_func'), that are given to a converter that manipulates your data and modifies it in place ... Using a class is probably a matter of personal preferences (would have done it using tool functions as I don't think there is much added value using a complete class); but using callables hiding odoo calls and modifying the input is a bit meh to me. Especially it is used once in the code: general rule is to keep code simple and local, and extract it in a more generic way if used at least 3 times.

Don't have much proposals to do: I don't really like the whole approach of using a class. My opinion is that it tries to do smart code, while all we need is a small moulinette processing nodes on a small tool inside the link tracker module.

Comment (tde-banana-odoo) [reply addons/link_tracker/tools/html_link_label_converter.py:11]: Without going into detailed review (as the main point is which kind of implementation we are going to do), I would say that I would prefer something that 1/ finds the URLs (allowing to early find duplicates if any) 2/ creates link tacker in batch 3/ updates them. Current implementation is very "generic" (black box approach) while I expected something very "odoo-ish" (tailored for the need and to be improved when necessary), as anyway we use it once currently. Probably yet another big debate to have.

File: addons/link_tracker/models/mail_render_mixin.py [hunk 12]
```diff
@@ -34,26 +35,28 @@ def _shorten_links(self, html, link_tracker_vals, blacklist=None, base_url=None)
...
             # Don't shorten urls present in blacklist (aka to skip list)
-            if blacklist and any(s in long_url for s in blacklist):
-                continue
-            label = (match[3] or '').strip()
+            if blacklist and any(s in url for s in blacklist):
+                return False
+            return True
 
-            create_vals = dict(link_tracker_vals, url=unescape(long_url), label=unescape(label))
+        def get_replacement_url(url, label):
```
Comment (tde-banana-odoo) [inline addons/link_tracker/models/mail_render_mixin.py:53]: Don't like that, especially you give an inner method called "get_replacement_url" that actually creates link trackers ... why adding complexity layers ?

File: addons/link_tracker/models/mail_render_mixin.py [hunk 13]
```diff
@@ -34,26 +35,28 @@ def _shorten_links(self, html, link_tracker_vals, blacklist=None, base_url=None)
...
+            return html
         base_url = base_url or self.env['ir.config_parameter'].sudo().get_param('web.base.url')
         short_schema = base_url + '/r/'
-        for match in set(re.findall(tools.HTML_TAG_URL_REGEX, html)):
-            long_url = match[1]
-            # Don't shorten already-shortened links
-            if long_url.startswith(short_schema):
-                continue
+
+        def should_convert_to_tracker(url):
```
Comment (tde-banana-odoo) [inline addons/link_tracker/models/mail_render_mixin.py:43]: Maybe we could define the inverse, like ``skip_convert``: shorter and more to the point ?

**Review by tde-banana-odoo — COMMENTED** [review]
Thanks for the PR, finally take some time to read it !

Well, have to say I am still not a big fan of the approach :boar: . The previous was doing an extensive usage of regex, this one is trying the black box approach that I don't really like (code harder to follow, extra layers, magic of input / output management, ...). Added some suggestions and questions, feel free to come back for discussion :) 

I understand you made a tool class to extract the tracker creation and try to hide the complexity, and in my opinion it is actually adding more complexity than removing, as we have to understand another custom class with callable magic. So, voilà, added some comments.

Cheers !

File: addons/link_tracker/tools/html_link_label_converter.py (hunk 11 above)
Comment (flch-odoo) [reply addons/link_tracker/tools/html_link_label_converter.py]: I moved the code in the function body, and started implementing a required `search_or_create_mutli`.
Shall I continue with this version or are there structural problems here as well?

File: addons/link_tracker/models/mail_render_mixin.py (hunk 12 above)
Comment (flch-odoo) [reply addons/link_tracker/models/mail_render_mixin.py]: This was expressly code that the caller passes, so they know what happens, but I agree that it requires readers to check another file if curious about the implementation.

**Review by flch-odoo — COMMENTED** [review]
Thank you @tde-banana-odoo ! Do you like this version better :) ?

File: addons/link_tracker/models/link_tracker.py [hunk 14]
```diff
@@ -126,28 +126,27 @@ def _check_unicity(self):
...
+        # Can not be implemented with a SQL constraint because we care about null values.
+        potential_duplicates = self.search(search_query)
+        duplicates = []
+        seen = set()
+        for tracker in potential_duplicates:
+            unique_fields = (tracker.url, tracker.campaign_id, tracker.medium_id, tracker.source_id, tracker.label)
+            if unique_fields in seen or seen.add(unique_fields):
+                duplicates.append(tracker)
+        if duplicates:
+            error_lines = '\n- '.join([
```
Comment (tde-banana-odoo) [inline addons/link_tracker/models/link_tracker.py:143]: Is the list mark necessary (can't we join an interator) ?

File: addons/link_tracker/models/link_tracker.py [hunk 15]
```diff
@@ -126,28 +126,27 @@ def _check_unicity(self):
...
+                str((tracker.url, tracker.campaign_id.name, tracker.medium_id.name, tracker.source_id.name, tracker.label))
+                for tracker in duplicates
+            ])
+            raise UserError(
+                _('The combination of Link Tracker values (URL, campaign, medium, source, and label) must be unique.\n'
+                  'The following combinations were already used: %(error_lines)s', error_lines=error_lines))
```
Comment (tde-banana-odoo) [inline addons/link_tracker/models/link_tracker.py:149]: Still voting for are :D 
```suggestion
                  'The following combinations are already used: %(error_lines)s', error_lines=error_lines))
```

File: addons/link_tracker/models/link_tracker.py [hunk 16]
```diff
@@ -126,28 +126,27 @@ def _check_unicity(self):
...
+            unique_fields = (tracker.url, tracker.campaign_id, tracker.medium_id, tracker.source_id, tracker.label)
+            if unique_fields in seen or seen.add(unique_fields):
+                duplicates.append(tracker)
+        if duplicates:
+            error_lines = '\n- '.join([
+                str((tracker.url, tracker.campaign_id.name, tracker.medium_id.name, tracker.source_id.name, tracker.label))
+                for tracker in duplicates
+            ])
+            raise UserError(
+                _('The combination of Link Tracker values (URL, campaign, medium, source, and label) must be unique.\n'
```
Comment (tde-banana-odoo) [inline addons/link_tracker/models/link_tracker.py:148]: Would remove the "the" that is a french tic, maybe we could use plural form ?
```suggestion
                _('Combinations of Link Tracker values (URL, campaign, medium, source, and label) must be unique.\n'
```

File: addons/link_tracker/models/link_tracker.py [hunk 17]
```diff
@@ -197,6 +196,27 @@ def search_or_create(self, vals):
...
+            (fname, '=', value)
+            for fname, value in vals.items()
+            if fname in {'url', 'campaign_id', 'medium_id', 'source_id', 'label'}
+        ] for vals in vals_list]
+
+        search_domain = expression.OR(search_domains)
+        result = self.search(search_domain)
+
+        if not result:
+            return self.create(vals_list)
```
Comment (tde-banana-odoo) [inline addons/link_tracker/models/link_tracker.py:216]: Technically you also have to check for duplicates inside ``vals_list``(a search or create is always more complicated in multi, this is why people often prefer the sequential one :D ) . You could have no existing trackers, but duplicates in the given list.

File: addons/link_tracker/models/link_tracker.py [hunk 18]
```diff
@@ -197,6 +196,27 @@ def search_or_create(self, vals):
...
+
+        raise NotImplemented(f"Found {len(result)} results: {result}")
```
Comment (tde-banana-odoo) [inline addons/link_tracker/models/link_tracker.py:218]: Can't you rebuild based on vals_list (from search or from create) ? A bit more complex but would allow a real multi search or create :)

File: addons/link_tracker/models/link_tracker.py [hunk 19]
```diff
@@ -197,6 +196,27 @@ def search_or_create(self, vals):
 
         return self.create(vals)
 
+    @api.model
+    def search_or_create_multi(self, vals_list):
```
Comment (tde-banana-odoo) [inline addons/link_tracker/models/link_tracker.py:200]: Could be great that 'search_or_create' calls 'search_or_create_multi' so that we centralize computation

File: addons/link_tracker/models/mail_render_mixin.py [hunk 20]
```diff
@@ -34,26 +38,72 @@ def _shorten_links(self, html, link_tracker_vals, blacklist=None, base_url=None)
...
-                continue
-            label = (match[3] or '').strip()
 
-            create_vals = dict(link_tracker_vals, url=unescape(long_url), label=unescape(label))
-            link = self.env['link.tracker'].search_or_create(create_vals)
-            if link.short_url:
-                # `str` manipulation required to support replacing "&" characters, common in urls
-                new_href = match[0].replace(long_url, link.short_url)
-                html = html.replace(markupsafe.Markup(match[0]), markupsafe.Markup(new_href))
+        root = lxml.html.fromstring(html)
```
Comment (tde-banana-odoo) [inline addons/link_tracker/models/mail_render_mixin.py:46]: Note that you may keep a ``tools/url_tracker.py`` file with the tool functions; there is a balance between imperative and class-based code :) (don't hate me) . I prefer small and clear tools to black box classes (that was my main point of the previous review).

**Review by tde-banana-odoo — COMMENTED** [review]
Zboing ! Such momentum on this one !

File: addons/link_tracker/models/link_tracker.py [hunk 21]
```diff
@@ -183,23 +188,69 @@ def create(self, vals_list):
...
-        result = self.search(search_domain, limit=1)
-
-        if result:
-            return result
-
-        return self.create(vals)
+        return self._search_or_create_multi([vals])
+
+    @api.model
+    def _search_or_create_multi(self, vals_list):
```
Comment (tde-banana-odoo) [inline addons/link_tracker/models/link_tracker.py:194]: Thanks for doing the extra code mile :D ! Well have to say it is not crystal clear to understand ... there are indeed several use cases (create with duplicates -> try to find them back when recreating the record set to return, merge with existing one in DB, ...).

Isn't it possible to use two structures to reduce the almost-n² in _match_records_with_vals and remove loops ? Like having 'tracker_key_to_idx' (tracker tuple -> indexes in vals_list, like unique_vals_to_vals_list_index), 'tracker_key_to_id' (tracker tuple -> existing id in DB), then build new_tracker_keys, update tracker_key_to_id, and rebuild final recordset based on tracker_key_to_id ?

It looks like what you are doing (there are not 32 ways to do it :D ) but it seems we could try to make the code a bit more fluid. What do you think ?

File: addons/link_tracker/models/mail_render_mixin.py [hunk 22]
```diff
@@ -86,3 +91,60 @@ def _shorten_links_text(self, content, link_tracker_vals, blacklist=None, base_u
                 content = re.sub(re.escape(original_url) + r'(?![\w@:%.+&~#=/-])', link.short_url, content)
 
         return content
+
+    @api.model
+    def _prepare_link_tracker_conversion(self, root_node, base_url, short_schema, blacklist):
```
Comment (tde-banana-odoo) [inline addons/link_tracker/models/mail_render_mixin.py:96]: Can't we make it a tool somewhere ? Not sure it belongs to the model (it essentially runs on a lxml node so seems a tool could be used here).

**Review by tde-banana-odoo — COMMENTED** [review]
Zboing, thanks for update ! Added some comments to polish the code and try to make it easier to read :) tell me what you think.

Cheers !

**Review by flch-odoo — COMMENTED** [review]
Thank you for this new pass, @tde-banana-odoo!

First answers "while the iron is hot" before I jump into optimization, so we can agree on the rest :)

File: addons/link_tracker/models/link_tracker.py [hunk 23]
```diff
@@ -46,6 +47,8 @@ class LinkTracker(models.Model):
     medium_id = fields.Many2one(ondelete='set null')
     source_id = fields.Many2one(ondelete='set null')
 
+    _sql_constraints = [("label_is_null_or_not_empty", "CHECK ( label <> '' )", "empty label should be NULL")]
```
[the remark this answers is not in the document]
Comment (flch-odoo) [reply addons/link_tracker/models/link_tracker.py:50]: Sorry I didn't link it on the task, it is now done.

File: addons/link_tracker/models/link_tracker.py (hunk 21 above)
Comment (flch-odoo) [reply addons/link_tracker/models/link_tracker.py:194]: I made another pass as these methods and found I could simplify it a lot while keeping the same structure.
Do you think it has become more readable even though we are assigning records in place immediately instead of handling another map?

Performance wise, without building a tree of record values to find matches faster, I don't think we can do better (except perhaps in the individual record matching logic). It does take O(m log n) + O (p log p) with m= unique vals_list items, n= nb existing records and p = nb unique non existing vals (= nb records to create).

**Review by flch-odoo — COMMENTED** [review]
The rest :)

File: addons/link_tracker/models/link_tracker.py (hunk 21 above)
Comment (tde-banana-odoo) [reply addons/link_tracker/models/link_tracker.py:194]: Had troubles understanding the computation, so in the end I pushed a wip commit with a proposal that I find easier to read: fetch existing, find new unique vals to create, then rebuild the final list. Did not run the whole test suit and pushed as a separate commit so that you can have a look and say what you think.

**Review by tde-banana-odoo — COMMENTED** [review]
Zboing, finally some news ! Ended up pushing a wip for the multi search or create as it was difficult to explain my pov. I let you say what you think of it :) .

**Review by flch-odoo — COMMENTED** [review]
Thank you for the code suggestion (you're now co-authoring :) ) , @tde-banana-odoo !

I've condensed it a little, I hope it is still readable enough.

I do not like having to handle in different places and a low-level that `""` and `False` are equal the case of our labels, (as it won't be as obvious as a SQL constraint and could cause confusion in 3rd party code) but I've made the change and will not fight for it at this stage.

Please also have a look at the new "link_tracker.tool.html" module where I simplified the now slightly generalized function to not handle duplicates as we support them in `search_or_create` (for which I added a deprecation notice for usage with `vals`, so we could drop support in 18.0 for example).

FYI, I think I'll still try to add a test for checking the Markup support.

File: addons/link_tracker/tools/html.py [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+def find_links_with_urls_and_labels(root_node, base_url, skip_regex=None, skip_prefix=None, skip_list=None):
+    """Return lxml link nodes and respective matching absolute urls and labels found in `root_node`.
+
+    :param lxml.etree._Element root_node: The root node to process
+    :param str base_url: base url to prefix relative hrefs
+    :param str skip_regex: URL pattern to skip
+    :param str skip_prefix: str prefix to skip
+    :param Iterable[str] skip_list: URLS to skip
+
+    :rtype: tuple[list[lxml.etree._Element], list[tuple[str, str]]]
```
Comment (tde-banana-odoo) [inline addons/link_tracker/tools/html.py:20]: Could return an Ordereddict, keys = node, value = {'url': absolute_url, 'label': label} maybe ? Easier to manipulate than a tuple whose second item is also a tuple

**Review by tde-banana-odoo — COMMENTED** [review]
Zblip, LGTM, just one proposal on tools :)

File: addons/link_tracker/models/mail_render_mixin.py [hunk 25]
```diff
@@ -34,29 +37,30 @@ def _shorten_links(self, html, link_tracker_vals, blacklist=None, base_url=None)
...
-                html = html.replace(markupsafe.Markup(match[0]), markupsafe.Markup(new_href))
+        root_node = lxml.html.fromstring(html)
+        link_nodes, urls_and_labels = find_links_with_urls_and_labels(
+            root_node, base_url, skip_regex=rf'^{URL_SKIP_PROTOCOL_REGEX}', skip_prefix=short_schema,
+            skip_list=blacklist)
+        if not link_nodes:
+            return html
+
+        links_trackers = self.env['link.tracker'].search_or_create([
+            dict(link_tracker_vals, url=unescape(url), label=unescape(label))
```
Comment (tde-banana-odoo) [inline addons/link_tracker/models/mail_render_mixin.py:53]: Actually do we have to unescape labels ? Isn't it a char title anyway ?

Comment (tde-banana-odoo) [inline addons/link_tracker/models/mail_render_mixin.py:53]: Shoudln't we unescape before calling "search_or_create" ? Unsure it is really the role of this "crud-like" method to do it

**Review by tde-banana-odoo — COMMENTED** [review]
Went through security warnings, two small questions :) (other points seems ok)

File: addons/link_tracker/models/mail_render_mixin.py (hunk 25 above)
Comment (flch-odoo) [reply addons/link_tracker/models/mail_render_mixin.py]: Turns out `lxml` unescapes both for us, so I could remove both calls (and went back to a dictionary^^).
I've added a test item for this behavior.

**Comment by tde-banana-odoo** [comment]
@robodoo override=ci/security

Reason
* 'unescape' import is just moved
* Markup is used to return Markup when input is Markup. What has been changed is that urls are replaced by short urls which means this replace is controlled by us; otherwise it is just a moulinette with html in / html out
* Markup used in tests to test Markup support
